File: src/main/java/degrel/visualize/GraphicsContextWrapper.java
package degrel.visualize;

import javafx.geometry.VPos;
import javafx.scene.canvas.Canvas;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.effect.BlendMode;
import javafx.scene.effect.Effect;
import javafx.scene.image.Image;
import javafx.scene.image.PixelWriter;
import javafx.scene.paint.Paint;
import javafx.scene.shape.ArcType;
import javafx.scene.shape.FillRule;
import javafx.scene.shape.StrokeLineCap;
import javafx.scene.shape.StrokeLineJoin;
import javafx.scene.text.Font;
import javafx.scene.text.TextAlignment;
import javafx.scene.transform.Affine;

/**
 * Decorator for javafx.scene.canvas.GraphicsContext
 */
public class GraphicsContextWrapper {
    private GraphicsContext origin;

    public Canvas getCanvas() {
        return origin.getCanvas();
    }

    public boolean isPointInPath(double x, double y) {
        return origin.isPointInPath(x, y);
    }

    public void appendSVGPath(String svgpath) {
        origin.appendSVGPath(svgpath);
    }

    public void drawImage(Image img, double sx, double sy, double sw, double sh, double dx, double dy, double dw, double dh) {
        origin.drawImage(img, sx, sy, sw, sh, dx, dy, dw, dh);
    }

    public Paint getFill() {
        return origin.getFill();
    }

    public void translate(double x, double y) {
        origin.translate(x, y);
    }

    public void translate(Vec v) {
        origin.translate(v.x(), v.y());
    }

    public void arc(double centerX, double centerY, double radiusX, double radiusY, double startAngle, double length) {
        origin.arc(centerX, centerY, radiusX, radiusY, startAngle, length);
    }

    public void closePath() {
        origin.closePath();
    }

    public void fillRect(double x, double y, double w, double h) {
        origin.fillRect(x, y, w, h);
    }

    public void setFillRule(FillRule fillRule) {
        origin.setFillRule(fillRule);
    }

    public void strokeOval(double x, double y, double w, double h) {
        origin.strokeOval(x, y, w, h);
    }

    public void strokeOval(Vec topLeft, double w, double h) {
        origin.strokeOval(topLeft.x(), topLeft.y(), w, h);
    }

    public void strokeOvalCenter(Vec center, double w, double h) {
        origin.strokeOval(center.x() - w / 2, center.y() - h / 2, w, h);
    }

    public void beginPath() {
        origin.beginPath();
    }

    public void setEffect(Effect e) {
        origin.setEffect(e);
    }

    public void drawImage(Image img, double x, double y) {
        origin.drawImage(img, x, y);
    }

    public void fillPolygon(double[] xPoints, double[] yPoints, int nPoints) {
        origin.fillPolygon(xPoints, yPoints, nPoints);
    }

    public void fillRoundRect(double x, double y, double w, double h, double arcWidth, double arcHeight) {
        origin.fillRoundRect(x, y, w, h, arcWidth, arcHeight);
    }

    public void stroke() {
        origin.stroke();
    }

    public StrokeLineCap getLineCap() {
        return origin.getLineCap();
    }

    public VPos getTextBaseline() {
        return origin.getTextBaseline();
    }

    public void strokePolyline(double[] xPoints, double[] yPoints, int nPoints) {
        origin.strokePolyline(xPoints, yPoints, nPoints);
    }

    public Paint getStroke() {
        return origin.getStroke();
    }

    public void drawImage(Image img, double x, double y, double w, double h) {
        origin.drawImage(img, x, y, w, h);
    }

    public void setTextAlign(TextAlignment align) {
        origin.setTextAlign(align);
    }

    public void lineTo(double x1, double y1) {
        origin.lineTo(x1, y1);
    }

    public void transform(double mxx, double myx, double mxy, double myy, double mxt, double myt) {
        origin.transform(mxx, myx, mxy, myy, mxt, myt);
    }

    public void setTransform(Affine xform) {
        origin.setTransform(xform);
    }

    public void save() {
        origin.save();
    }

    public void setFont(Font f) {
        origin.setFont(f);
    }

    public void rotate(double degrees) {
        origin.rotate(degrees);
    }

    public void moveTo(double x0, double y0) {
        origin.moveTo(x0, y0);
    }

    public void setLineWidth(double lw) {
        origin.setLineWidth(lw);
    }

    public void bezierCurveTo(double xc1, double yc1, double xc2, double yc2, double x1, double y1) {
        origin.bezierCurveTo(xc1, yc1, xc2, yc2, x1, y1);
    }

    public FillRule getFillRule() {
        return origin.getFillRule();
    }

    public void arcTo(double x1, double y1, double x2, double y2, double radius) {
        origin.arcTo(x1, y1, x2, y2, radius);
    }

    public Effect getEffect(Effect e) {
        return origin.getEffect(e);
    }

    public double getMiterLimit() {
        return origin.getMiterLimit();
    }

    public void applyEffect(Effect e) {
        origin.applyEffect(e);
    }

    public void setGlobalAlpha(double alpha) {
        origin.setGlobalAlpha(alpha);
    }

    public void restore() {
        origin.restore();
    }

    public void setTextBaseline(VPos baseline) {
        origin.setTextBaseline(baseline);
    }

    public void setFill(Paint p) {
        origin.setFill(p);
    }

    public void clearRect(double x, double y, double w, double h) {
        origin.clearRect(x, y, w, h);
    }

    public void strokeRect(double x, double y, double w, double h) {
        origin.strokeRect(x, y, w, h);
    }

    public void strokeArc(double x, double y, double w, double h, double startAngle, double arcExtent, ArcType closure) {
        origin.strokeArc(x, y, w, h, startAngle, arcExtent, closure);
    }

    public void rect(double x, double y, double w, double h) {
        origin.rect(x, y, w, h);
    }

    public void setStroke(Paint p) {
        origin.setStroke(p);
    }

    public void fill() {
        origin.fill();
    }

    public BlendMode getGlobalBlendMode() {
        return origin.getGlobalBlendMode();
    }

    public void setLineJoin(StrokeLineJoin join) {
        origin.setLineJoin(join);
    }

    public void setGlobalBlendMode(BlendMode op) {
        origin.setGlobalBlendMode(op);
    }

    public void setLineCap(StrokeLineCap cap) {
        origin.setLineCap(cap);
    }

    public void transform(Affine xform) {
        origin.transform(xform);
    }

    public void setMiterLimit(double ml) {
        origin.setMiterLimit(ml);
    }

    public void strokeRoundRect(double x, double y, double w, double h, double arcWidth, double arcHeight) {
        origin.strokeRoundRect(x, y, w, h, arcWidth, arcHeight);
    }

    public void strokeText(String text, double x, double y, double maxWidth) {
        origin.strokeText(text, x, y, maxWidth);
    }

    public void quadraticCurveTo(double xc, double yc, double x1, double y1) {
        origin.quadraticCurveTo(xc, yc, x1, y1);
    }

    public PixelWriter getPixelWriter() {
        return origin.getPixelWriter();
    }

    public Font getFont() {
        return origin.getFont();
    }

    public void strokeLine(double x1, double y1, double x2, double y2) {
        origin.strokeLine(x1, y1, x2, y2);
    }

    public void strokeLine(Vec v1, Vec v2) {
        origin.strokeLine(v1.x(), v1.y(), v2.x(), v2.y());
    }

    public void setTransform(double mxx, double myx, double mxy, double myy, double mxt, double myt) {
        origin.setTransform(mxx, myx, mxy, myy, mxt, myt);
    }

    public void fillText(String text, double x, double y, double maxWidth) {
        origin.fillText(text, x, y, maxWidth);
    }

    public void fillOval(double x, double y, double w, double h) {
        origin.fillOval(x, y, w, h);
    }

    public void fillOval(Vec topLeft, double w, double h) {
        origin.fillOval(topLeft.x(), topLeft.y(), w, h);
    }

    public void fillOvalCenter(Vec center, double w, double h) {
        origin.fillOval(center.x() - w / 2, center.y() - h / 2, w, h);
    }

    public void fillArc(double x, double y, double w, double h, double startAngle, double arcExtent, ArcType closure) {
        origin.fillArc(x, y, w, h, startAngle, arcExtent, closure);
    }

    public TextAlignment getTextAlign() {
        return origin.getTextAlign();
    }

    public Affine getTransform() {
        return origin.getTransform();
    }

    public void strokePolygon(double[] xPoints, double[] yPoints, int nPoints) {
        origin.strokePolygon(xPoints, yPoints, nPoints);
    }

    public double getGlobalAlpha() {
        return origin.getGlobalAlpha();
    }

    public Affine getTransform(Affine xform) {
        return origin.getTransform(xform);
    }

    public StrokeLineJoin getLineJoin() {
        return origin.getLineJoin();
    }

    public double getLineWidth() {
        return origin.getLineWidth();
    }

    public void strokeText(String text, double x, double y) {
        origin.strokeText(text, x, y);
    }

    public void scale(double x, double y) {
        origin.scale(x, y);
    }

    public void clip() {
        origin.clip();
    }

    public void fillText(String text, double x, double y) {
        origin.fillText(text, x, y);
    }

    public GraphicsContextWrapper(GraphicsContext origin) {
        this.origin = origin;
    }
}

File: src/main/scala/degrel/ui/GraphicsContextWrapper.scala
package degrel.ui

import javafx.geometry.VPos
import javafx.scene.canvas.{Canvas, GraphicsContext}
import javafx.scene.effect.{BlendMode, Effect}
import javafx.scene.image.{Image, PixelWriter}
import javafx.scene.paint.Paint
import javafx.scene.shape.{ArcType, FillRule, StrokeLineCap, StrokeLineJoin}
import javafx.scene.text.{Font, TextAlignment}
import javafx.scene.transform.Affine

import degrel.ui.Vec

/**
 * Decorator for javafx.scene.canvas.GraphicsContext
 */
class GraphicsContextWrapper(val origin: GraphicsContext) {
  def getCanvas: Canvas = {
    origin.getCanvas
  }

  def isPointInPath(x: Double, y: Double): Boolean = {
    origin.isPointInPath(x, y)
  }

  def appendSVGPath(svgpath: String) = {
    origin.appendSVGPath(svgpath)
  }

  def drawImage(img: Image,
                sx: Double,
                sy: Double,
                sw: Double,
                sh: Double,
                dx: Double,
                dy: Double,
                dw: Double,
                dh: Double) = {
    origin.drawImage(img, sx, sy, sw, sh, dx, dy, dw, dh)
  }

  def getFill: Paint = {
    origin.getFill
  }

  def translate(x: Double, y: Double) = {
    origin.translate(x, y)
  }

  def translate(v: Vec) = {
    origin.translate(v.x, v.y)
  }

  def arc(centerX: Double,
          centerY: Double,
          radiusX: Double,
          radiusY: Double,
          startAngle: Double,
          length: Double) = {
    origin.arc(centerX, centerY, radiusX, radiusY, startAngle, length)
  }

  def closePath() = {
    origin.closePath()
  }

  def fillRect(x: Double, y: Double, w: Double, h: Double) = {
    origin.fillRect(x, y, w, h)
  }

  def setFillRule(fillRule: FillRule) = origin.setFillRule(fillRule)

  def strokeOval(x: Double, y: Double, w: Double, h: Double) = {
    origin.strokeOval(x, y, w, h)
  }

  def strokeOval(topLeft: Vec, w: Double, h: Double) = {
    origin.strokeOval(topLeft.x, topLeft.y, w, h)
  }

  def strokeOvalCenter(center: Vec, w: Double, h: Double) = {
    origin.strokeOval(center.x - w / 2, center.y - h / 2, w, h)
  }

  def beginPath() = {
    origin.beginPath()
  }

  def setEffect(e: Effect) = {
    origin.setEffect(e)
  }

  def drawImage(img: Image, x: Double, y: Double) = {
    origin.drawImage(img, x, y)
  }

  def fillPolygon(xPoints: Array[Double], yPoints: Array[Double], nPoints: Int) = {
    origin.fillPolygon(xPoints, yPoints, nPoints)
  }

  def fillRoundRect(x: Double, y: Double, w: Double, h: Double, arcWidth: Double, arcHeight: Double) = {
    origin.fillRoundRect(x, y, w, h, arcWidth, arcHeight)
  }

  def stroke() = {
    origin.stroke()
  }

  def getLineCap: StrokeLineCap = {
    origin.getLineCap
  }

  def getTextBaseline: VPos = {
    origin.getTextBaseline
  }

  def strokePolyline(xPoints: Array[Double], yPoints: Array[Double], nPoints: Int) = {
    origin.strokePolyline(xPoints, yPoints, nPoints)
  }

  def getStroke: Paint = {
    origin.getStroke
  }

  def drawImage(img: Image, x: Double, y: Double, w: Double, h: Double) = {
    origin.drawImage(img, x, y, w, h)
  }

  def setTextAlign(align: TextAlignment) = {
    origin.setTextAlign(align)
  }

  def lineTo(x1: Double, y1: Double) = {
    origin.lineTo(x1, y1)
  }

  def transform(mxx: Double, myx: Double, mxy: Double, myy: Double, mxt: Double, myt: Double) = {
    origin.transform(mxx, myx, mxy, myy, mxt, myt)
  }

  def setTransform(xform: Affine) = {
    origin.setTransform(xform)
  }

  def save() = {
    origin.save()
  }

  def setFont(f: Font) = {
    origin.setFont(f)
  }

  def rotate(degrees: Double) = {
    origin.rotate(degrees)
  }

  def moveTo(x0: Double, y0: Double) = {
    origin.moveTo(x0, y0)
  }

  def setLineWidth(lw: Double) = {
    origin.setLineWidth(lw)
  }

  def bezierCurveTo(xc1: Double, yc1: Double, xc2: Double, yc2: Double, x1: Double, y1: Double) = {
    origin.bezierCurveTo(xc1, yc1, xc2, yc2, x1, y1)
  }

  def getFillRule: FillRule = {
    origin.getFillRule
  }

  def arcTo(x1: Double, y1: Double, x2: Double, y2: Double, radius: Double) = {
    origin.arcTo(x1, y1, x2, y2, radius)
  }

  def getEffect(e: Effect): Effect = {
    origin.getEffect(e)
  }

  def getMiterLimit: Double = {
    origin.getMiterLimit
  }

  def applyEffect(e: Effect) = {
    origin.applyEffect(e)
  }

  def setGlobalAlpha(alpha: Double) = {
    origin.setGlobalAlpha(alpha)
  }

  def restore() = {
    origin.restore()
  }

  def setTextBaseline(baseline: VPos) = {
    origin.setTextBaseline(baseline)
  }

  def setFill(p: Paint) = {
    origin.setFill(p)
  }

  def clearRect(x: Double, y: Double, w: Double, h: Double) = {
    origin.clearRect(x, y, w, h)
  }

  def strokeRect(x: Double, y: Double, w: Double, h: Double) = {
    origin.strokeRect(x, y, w, h)
  }

  def strokeArc(x: Double, y: Double, w: Double, h: Double, startAngle: Double, arcExtent: Double, closure: ArcType) = {
    origin.strokeArc(x, y, w, h, startAngle, arcExtent, closure)
  }

  def rect(x: Double, y: Double, w: Double, h: Double) = {
    origin.rect(x, y, w, h)
  }

  def setStroke(p: Paint) = {
    origin.setStroke(p)
  }

  def fill() = {
    origin.fill()
  }

  def getGlobalBlendMode: BlendMode = {
    origin.getGlobalBlendMode
  }

  def setLineJoin(join: StrokeLineJoin) = {
    origin.setLineJoin(join)
  }

  def setGlobalBlendMode(op: BlendMode) = {
    origin.setGlobalBlendMode(op)
  }

  def setLineCap(cap: StrokeLineCap) = {
    origin.setLineCap(cap)
  }

  def transform(xform: Affine) = {
    origin.transform(xform)
  }

  def setMiterLimit(ml: Double) = {
    origin.setMiterLimit(ml)
  }

  def strokeRoundRect(x: Double, y: Double, w: Double, h: Double, arcWidth: Double, arcHeight: Double) = {
    origin.strokeRoundRect(x, y, w, h, arcWidth, arcHeight)
  }

  def strokeText(text: String, x: Double, y: Double, maxWidth: Double) = {
    origin.strokeText(text, x, y, maxWidth)
  }

  def quadraticCurveTo(xc: Double, yc: Double, x1: Double, y1: Double) = {
    origin.quadraticCurveTo(xc, yc, x1, y1)
  }

  def getPixelWriter: PixelWriter = {
    origin.getPixelWriter
  }

  def getFont: Font = {
    origin.getFont
  }

  def strokeLine(x1: Double, y1: Double, x2: Double, y2: Double) = {
    origin.strokeLine(x1, y1, x2, y2)
  }

  def strokeLine(v1: Vec, v2: Vec) = {
    origin.strokeLine(v1.x, v1.y, v2.x, v2.y)
  }

  def setTransform(mxx: Double, myx: Double, mxy: Double, myy: Double, mxt: Double, myt: Double) = {
    origin.setTransform(mxx, myx, mxy, myy, mxt, myt)
  }

  def fillText(text: String, x: Double, y: Double, maxWidth: Double) = {
    origin.fillText(text, x, y, maxWidth)
  }

  def fillOval(x: Double, y: Double, w: Double, h: Double) = {
    origin.fillOval(x, y, w, h)
  }

  def fillOval(topLeft: Vec, w: Double, h: Double) = {
    origin.fillOval(topLeft.x, topLeft.y, w, h)
  }

  def fillOvalCenter(center: Vec, w: Double, h: Double) = {
    origin.fillOval(center.x - w / 2, center.y - h / 2, w, h)
  }

  def fillOvalCenter(x: Double, y: Double, w: Double, h: Double) = {
    origin.fillOval(x - w / 2, y - h / 2, w, h)
  }

  def fillArc(x: Double, y: Double, w: Double, h: Double, startAngle: Double, arcExtent: Double, closure: ArcType) = {
    origin.fillArc(x, y, w, h, startAngle, arcExtent, closure)
  }

  def getTextAlign: TextAlignment = {
    origin.getTextAlign
  }

  def getTransform: Affine = {
    origin.getTransform
  }

  def strokePolygon(xPoints: Array[Double], yPoints: Array[Double], nPoints: Int) = {
    origin.strokePolygon(xPoints, yPoints, nPoints)
  }

  def getGlobalAlpha: Double = {
    origin.getGlobalAlpha
  }

  def getTransform(xform: Affine): Affine = {
    origin.getTransform(xform)
  }

  def getLineJoin: StrokeLineJoin = {
    origin.getLineJoin
  }

  def getLineWidth: Double = {
    origin.getLineWidth
  }

  def strokeText(text: String, x: Double, y: Double) = {
    origin.strokeText(text, x, y)
  }

  def scale(x: Double, y: Double) = {
    origin.scale(x, y)
  }

  def clip() = {
    origin.clip()
  }

  def fillText(text: String, x: Double, y: Double) = {
    origin.fillText(text, x, y)
  }

  /**
   * 矢印を描画します．
   * @param from 矢印の根元の座標
   * @param to 矢印の先の座標．鏃が描画(fill)されます
   * @param arrowHeight 鏃の長さ(高さ) 矢の方向の鏃の大きさです
   * @param arrowWidth 鏃の幅 矢の方向に垂直な方向の大きさです．この値は矢の中心からの半径に相当する長さで，鏃自体の幅はこれの2倍になります．
   */
  def fillArrow(from: Vec, to: Vec, arrowHeight: Double = 10, arrowWidth: Double = 5) = {
    origin.strokeLine(from.x, from.y, to.x, to.y)
    // 筈から鏃までのベクトル
    val delta = to - from
    // 鏃とシャフトの部分の交点の座標
    val base = to - delta.normalize(arrowHeight)
    // 鏃の三角形の筈側2点を求めます．シャフト方向のベクトルをPI/2回転させて，arrowWidthで長さを調整します
    val side = delta.rotate(Math.PI / 2).normalize(arrowWidth)
    val wing1 = base + side
    val wing2 = base - side
    origin.beginPath()
    origin.moveTo(to.x, to.y)
    origin.lineTo(wing1.x, wing1.y)
    origin.lineTo(wing2.x, wing2.y)
    origin.closePath()
    origin.fill()
  }
}